### PR TITLE
R21.1 format cura lists

### LINKED ISSUE #91: Mod mesh Top Bottom Line Directions, Infill Line Directions job error.
Setting the default empty list for Top Bottom Line Directions or Infill Line Directions to a mod mesh will cause a job error.

Explicitly defining the list for Top Bottom Line Directions as [45, 135] however, doesn't cause a job error.

### PR DESCRIPTION
**Pull request by AijunHall**
Should fix #91 

When users in the plugin enter values like `[ ]` or `[45, 135, ]` for the skin angles and infill angles, `from_cura_setting` was causing a crash from attempting to parse empty string values, so I'm trying to consolidate the string-to-list formatting here.

### CODE REVIEW

File: pywim/am.py [hunk 1]
```diff
@@ -89,18 +98,21 @@ def from_cura_setting(self, name, value, set_auxiliary=True):
             # infill pattern and we could be in the middle of processing multiple cura
             # settings and the infill pattern could be "on deck" to be changed
         elif name == 'infill_angles':
-            angles = value.strip('[]').split(',')
+            angles = self._format_cura_list(value)
+            if not angles:
```
Comment (AijunHall) [inline pywim/am.py:102]: I'm not sure what should happen in the event of empty infill angle lists. I see Cura's default values here, but these would conflict with the validation of having a single infill orientation value.

image

Comment (AijunHall) [reply pywim/am.py:102]: I found the default Infill orientation is just `0.0` so is the return statement sufficient, or does the float value of 0.0 need to be specified?

https://github.com/tetonsim/pywim/blob/43840956b96b1f365df263e52ae93eeb89be8092/pywim/am.py#L27

Comment (bradymadams) [reply pywim/am.py:102]: The default of zero is actually inconsistent with Cura. I don't want to change it cause I don't know what other implications that might have. But for the purposes of these changes - if len(angles) is zero then we should set `self.infill.orientation = 45.0`.

Can you double check in Cura that if you explicitly set infill angles to an empty list that you get a 45 degree infill orientation? If not, I might be mistaken, so let me know what you observe.

**Review by bradymadams — CHANGES_REQUESTED** [review]

**Comment by AijunHall** [comment]
@bradymadams 
> Can you double check in Cura that if you explicitly set infill angles to an empty list that you get a 45 degree infill orientation? If not, I might be mistaken, so let me know what you observe.
According to Cura (in the screenshot above) and after looking back into the `fdmprinter.def.json`, this is what it has to say about default empty infill angles:

*"Default is an empty list which means use the traditional default angles (45 and 135 degrees for the lines and zig zag patterns and 45 degrees for all other patterns)."*

Currently, in the checkJob validation we set empty infill angle lists to 45 if thats what you mean by 
> if you explicitly set infill angles to an empty list that you get a 45 degree infill orientation

**Comment by bradymadams** [comment]
> Currently, in the plugin-side checkJob validation we set empty infill angle lists to 45 if thats what you mean by
> 
> > if you explicitly set infill angles to an empty list that you get a 45 degree infill orientation

I meant to explicitly set the property to `[]` (maybe for both normal and mod mesh) and slice and visually inspect how the infill is oriented. I suppose the settings in fdmprinter.def.json should back up what you observe.

Ultimately, if the user sets infill angles to `[]` we should set the infill orientation in the `Config` object to whatever is consistent with what Cura does.

**Comment by AijunHall** [comment]
> I meant to explicitly set the property to `[]` (maybe for both normal and mod mesh) and slice and visually inspect how the infill is oriented. I suppose the settings in fdmprinter.def.json should back up what you observe.

Oh I see now. The infill angles here seem to match up as expected

Default `[ ]`
image

Explicitly defined `[45]`
image

**Comment by AijunHall** [comment]
I understand now what should happen to the infill angles. I'm looking into why the 45 value isn't getting transferred to the config from the plugin validation, since the current return statement should be unnecessary in that case.

**Comment by AijunHall** [comment]
@bradymadams After some more debugging on my end, I'm not sure on how to proceed with this. The real issue here is that individual mesh print settings aren't being saved properly via the 3mf. 

For mod meshes, I'm running into situations where the mod mesh skin_angles/infill_angles work well with these suggested changes, but instead an empty default extruder skin_angles/infill_angles is being passed along and causing an Index Error crash (only for infill angles which is why I asked about the return statement) This probably falls under the scope of [561](https://github.com/tetonsim/is-cura-ui/issues/561) in is-cura-ui

I see some options here being:

- leave lines 102-103 as is to avoid an Index Error crash
- define the default `skin_orientations` and `infill.orientation`. Something like this:
```
elif name == "infill_angles":
    angles = self._format_cura_list(value)
    if not angles:
        self.infill.orientation = 45.0

elif name == "skin_angles":
    angles = self._format_cura_list(value)
    if not angles:
        self.skin_orientations = [45, 135]
```
- delete line 102-103 and keep this new Index Error crash

Option 1 and 2 rub me the wrong way, but I don't know what would be most appropriate here.

**Comment by bradymadams** [comment]
> but instead an empty default extruder skin_angles/infill_angles is being passed along and causing an Index Error crash

I'm not sure I understand fully, but regardless if some print settings are saved correctly or incorrectly, I think our code should be able to parse various settings, including empty lists.

> Option 1 and 2 rub me the wrong way since they seem like a quick and dirty patch to the root problem

Can you explain a little more? It doesn't seem quick and dirty to me. If an empty list means the infill orientation is 45 degrees, (and a similar rule for skin angles) then I think it's fully correct to detect an empty list and set those properties appropriately. But maybe I'm not understanding the entire problem.

**Comment by AijunHall** [comment]
> > Option 1 and 2 rub me the wrong way since they seem like a quick and dirty patch to the root problem
> 
> Can you explain a little more? It doesn't seem quick and dirty to me. If an empty list means the infill orientation is 45 degrees, (and a similar rule for skin angles) then I think it's fully correct to detect an empty list and set those properties appropriately. But maybe I'm not understanding the entire problem.

To me it doesn't seem like the purpose of `from_cura_setting()` is to parse empty lists to the Cura values, but I understand and agree with:

> if some print settings are saved correctly or incorrectly, I think our code should be able to parse various settings, including empty lists. 

This is really more of an issue with the property handling in is-cura-ui. I'll make the changes, I wanted to make sure it was appropriate to have this functionality within the scope of this ticket/PR if that clarifies my concerns.

**Comment by bradymadams** [comment]
> To me it doesn't seem like the purpose of `from_cura_setting()` is to parse empty lists to the Cura values since that should be handled beforehand in the propertyhandler/3mf saving, but I understand and agree with:

It may be the case that the 3MF saving in the plugin should (or will in the future) handle these cases and translate empty lists to non-empty lists. But robustness is an important part of software design. We can't guarantee every 3MF sent to the backend will be saved through the same logic. In an API we can't really make any assumption about where the data has come from. So we design our software to be robust and handle expected cases, and unexpected cases should throw appropriate errors*. I think it's definitely within the "expected cases" to get an empty list as that is a completely valid setting in Cura.

*This is the goal, but obviously we don't always meet this goal. Hence the need for improvements and bug fixes.

**Review by bradymadams — APPROVED** [review]

**Pull request merged**